consensus/poet/core/sawtooth_poet/poet_consensus/poet_block_publisher.py

```diff
@@ -321,6 +346,28 @@ def initialize_block(self, block_header):
                 'key to show up in validator registry.',
                 validator_info.signup_info.poet_public_key[:8],
                 validator_info.signup_info.poet_public_key[-8:])
+
+            # Check if we need to give up on this registration attempt
+            consensus_state = \
+                ConsensusState.consensus_state_for_block_id(
+                    block_id=block_header.previous_block_id,
+                    block_cache=self._block_cache,
+                    state_view_factory=self._state_view_factory,
+                    consensus_state_store=self._consensus_state_store,
+                    poet_enclave_module=poet_enclave_module)
+            poet_settings_view = PoetSettingsView(state_view)
+            signup_nonce = poet_key_state.signup_nonce
+            if consensus_state.signup_attempt_timed_out(signup_nonce,
+                                                        poet_settings_view,
+                                                        self._block_cache):
+                LOGGER.error('Assuming registration txn failed using PPK '
+                             '%s Retry registration',
+                             poet_key_state.poet_public_key)
+                del self._poet_key_state_store[poet_key_state.poet_public_key]
+                self._register_signup_information(
+                    block_header=block_header,
+                    poet_enclave_module=poet_enclave_module)
+
```

Comment:
Can you refactor this into a function since it looks mostly identical to the above on line 275? Also, the code to construct the consensus_state object is duplicated three times in this function. It should probably be moved to a separate function and we should make sure that consensus_state is only constructed once per call to this function at most, since it is probably expensive.